tests/check_public_api.py

```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/python3
+
+import subprocess
+import tempfile
+import os
+import re
+import sys
+
+
+def get_public_symbols_from_shared_object(prefix):
+    """
+    Find public (exported) symbols in a shared library (*.so) by calling
+    the 'nm' command.
+    :param prefix: CMAKE_INSTALL_PREFIX
+    :return: a set of public symbols
+    """
+    so_path = os.path.join(prefix, "lib64", "libopenscap.so")
+    nm_command = ["nm", "--demangle", "--dynamic", "--defined-only", so_path]
+    nm_output_bytes = subprocess.check_output(nm_command)
+    nm_output = nm_output_bytes.decode(encoding="utf-8")
+    symbols = {line.split()[2] for line in nm_output.splitlines()}
+    return symbols
+
+
+def get_public_headers(prefix):
+    """
+    It is useful to check if all the public headers are actually installed
+    under $PREFIX/include/openscap.
+    :param prefix: CMAKE_INSTALL_PREFIX
+    :return: a set of public headers filepaths
+    """
+    public_headers = set()
+    include_path = os.path.join(prefix, "include", "openscap")
+    for root, dirs, files in os.walk(include_path):
+        for f in files:
+            full_path = os.path.join(root, f)
+            public_headers.add(full_path)
+    return public_headers
+
+
+def get_public_symbols_from_header(header_path):
+    """
+    Parses a header file to find symbols marked by OSCAP_API macro.
+    :param header_path: path to the header file
+    :return: a list of public symbols in the header file
+    """
+    symbols = set()
+    prototype_regex = re.compile(r"^[ \t]*OSCAP_API[^;]*;", re.MULTILINE)
+    with open(header_path, "r") as header_file:
+        content = header_file.read()
+    prototypes = prototype_regex.findall(content)
+    function_name_regex = re.compile(r"([a-zA-Z0-9_]+)\s*\(.*\)")
+    for p in prototypes:
+        p = p.replace("\n", " ")
+        f_match = function_name_regex.search(p)
+        f_name = f_match.group(1) if f_match else "unknown"
+        symbols.add(f_name)
+    return symbols
+
+
+def get_public_symbols_from_headers(headers):
+    """
+    Find public (exported) symbols in header files by finding symbols
+    marked as OSCAP_API.
+    :param headers: a list of public headers file paths
+    :return: a set of public symbols in public headers
+    """
+    symbols = set()
+    for h in headers:
+        symbols.update(get_public_symbols_from_header(h))
+    return symbols
+
+
+def _run_command(command):
+    try:
+        subprocess.run(command, check=True,
+                       stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                       encoding="utf-8")
+    except subprocess.CalledProcessError as e:
+        print(e.output)
```

Comment:
I'd prefer if this merged `stdout` and `stderr` so that they were interleaved as during build `stderr` gets written to and having context of `stdout` helps (by setting `stderr=subprocess.STDOUT`), but I won't hold up merging this on that.